File: 2021/2021W50/preppindataW50.R
library(readxl)
library(dplyr)
library(tidyr)

xlsx <- "C:/Data/PreppinData/Sales Department Input.xlsx"

final <-  read_excel(xlsx, 'October', col_types = c("numeric", "date", "text", "numeric", "numeric", "numeric", "text", "numeric")) %>%
  rename('Oct YTD Total' = names(.)[length(names(.))]) %>%
  bind_rows(., read_excel(xlsx, 'November', col_types = c("numeric", "date", "text", "numeric", "numeric", "numeric", "text"))) %>%
  fill(`Salesperson`, .direction = 'up') %>%
  group_by(`Salesperson`) %>%
  mutate('Oct YTD Total' = max(`Oct YTD Total`, na.rm = TRUE),
         'Month' = format(`Date`,'%m')) %>%
  filter(!is.na(`Date`)) %>%
  pivot_longer(cols=c('Road', 'Gravel', 'Mountain'), names_to = 'Bike Type', values_to = 'Sales') %>%
  group_by(`Salesperson`, `Month`) %>%
  mutate('Monthly Total' = sum(`Sales`),
         'YTD Total' = `Oct YTD Total` + if_else(`Month`==11, `Monthly Total`, 0)) %>%
  ungroup() %>%
  select(c('Salesperson', 'Date', 'Bike Type', 'Sales', 'YTD Total'))
  
View(final)

File: 2021/2021W50/PreppindataW50.py
import pandas as pd

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\Sales Department Input.xlsx")

temp = []
for sheet in xlsx.sheet_names:
    df = pd.read_excel(xlsx,sheet)
    df.insert(0,'Month',sheet)
    temp.append(df)
sales = pd.concat(temp)

sales = sales.rename(columns={sales.columns[-1]:'Oct YTD Total'})
sales[['Salesperson','Oct YTD Total']] = sales[['Salesperson','Oct YTD Total']].fillna(method='bfill') 
sales = sales[pd.notna(sales['Date'])]

sales['Oct YTD Total'] = sales['Oct YTD Total'].groupby(sales['Salesperson']).transform('max')
sales['Monthly Total'] = sales['Total'].groupby([sales['Month'], sales['Salesperson']]).transform('sum')
sales['YTD Total'] = sales.apply(lambda x: x['Oct YTD Total'] + (x['Monthly Total'] if x['Month']=='November' else 0), axis=1)

sales = sales[['Date', 'Salesperson', 'Road', 'Gravel', 'Mountain', 'YTD Total']].copy()
sales = sales.melt(id_vars=['Salesperson', 'Date', 'YTD Total'], value_name='Sales', var_name='Bike Type')
sales = sales[['Salesperson', 'Date', 'Bike Type', 'Sales', 'YTD Total']]
